src/storage/v2/disk/label_index.cpp

```diff
@@ -0,0 +1,209 @@
+      if (!disk_transaction
+               ->Put(utils::SerializeVertexAsKeyForLabelIndex(index_label, vertex.gid),
+                     utils::SerializeVertexAsValueForLabelIndex(index_label, vertex.labels, vertex.properties))
+               .ok()) {
+        return false;
+      }
+    }
+  }
+
+  return CommitWithTimestamp(disk_transaction.get(), commit_timestamp);
+}
+
+/// TODO: this can probably be optimized
+bool DiskLabelIndex::ClearDeletedVertex(std::string_view gid, uint64_t transaction_commit_timestamp) const {
+  auto disk_transaction = CreateAllReadingRocksDBTransaction();
+
+  rocksdb::ReadOptions ro;
+  std::string strTs = utils::StringTimestamp(std::numeric_limits<uint64_t>::max());
+  rocksdb::Slice ts(strTs);
+  ro.timestamp = &ts;
+  auto it = std::unique_ptr<rocksdb::Iterator>(disk_transaction->GetIterator(ro));
+  for (it->SeekToFirst(); it->Valid(); it->Next()) {
+    if (std::string key = it->key().ToString(); gid == utils::ExtractGidFromLabelIndexStorage(key)) {
+      if (!disk_transaction->Delete(key).ok()) {
+        return false;
+      }
+    }
+  }
+
+  return CommitWithTimestamp(disk_transaction.get(), transaction_commit_timestamp);
+}
+
+bool DiskLabelIndex::DeleteVerticesWithRemovedIndexingLabel(uint64_t transaction_start_timestamp,
+                                                            uint64_t transaction_commit_timestamp) {
+  auto disk_transaction = CreateAllReadingRocksDBTransaction();
+
+  rocksdb::ReadOptions ro;
+  std::string strTs = utils::StringTimestamp(std::numeric_limits<uint64_t>::max());
+  rocksdb::Slice ts(strTs);
+  ro.timestamp = &ts;
+  bool deletion_success = true;
+  entries_for_deletion.WithLock([&deletion_success, transaction_start_timestamp,
+                                 disk_transaction_ptr = disk_transaction.get()](auto &tx_to_entries_for_deletion) {
+    if (auto tx_it = tx_to_entries_for_deletion.find(transaction_start_timestamp);
+        tx_it != tx_to_entries_for_deletion.end()) {
+      deletion_success = ClearTransactionEntriesWithRemovedIndexingLabel(*disk_transaction_ptr, tx_it->second);
+      tx_to_entries_for_deletion.erase(tx_it);
+    }
+  });
+  if (deletion_success) {
+    return CommitWithTimestamp(disk_transaction.get(), transaction_commit_timestamp);
+  }
+  return false;
+}
+
+void DiskLabelIndex::UpdateOnAddLabel(LabelId added_label, Vertex *vertex_before_update, const Transaction &tx) {
+  entries_for_deletion.WithLock([added_label, vertex_before_update, &tx](auto &tx_to_entries_for_deletion) {
+    if (auto tx_it = tx_to_entries_for_deletion.find(tx.start_timestamp); tx_it != tx_to_entries_for_deletion.end()) {
+      if (auto vertex_label_index_it = tx_it->second.find(vertex_before_update->gid);
+          vertex_label_index_it != tx_it->second.end()) {
+        std::erase_if(vertex_label_index_it->second,
+                      [added_label](const LabelId &indexed_label) { return indexed_label == added_label; });
+      }
+    }
```

Comment:
```suggestion
  auto tx_it = tx_to_entries_for_deletion.find(tx.start_timestamp);
    if (tx_it == tx_to_entries_for_deletion.end()) {
      return;
    }
    auto vertex_label_index_it = tx_it->second.find(vertex_before_update->gid);
    if (vertex_label_index_it != tx_it->second.end()) {
      return;
    }
    std::erase_if(vertex_label_index_it->second,
                  [added_label](const LabelId &indexed_label) { return indexed_label == added_label; });
      
```

I think this is a bit cleaner, as it has early returns where possible, and thus slightly less indents